components/eam/src/physics/cam/modal_aer_opt.F90

```diff
@@ -1553,71 +1551,114 @@ end subroutine modal_size_parameters
-          enddo
-   10     ix(ic)=max0(i-1,1)
-          ip1=min(ix(ic)+1,im)
-          dx=(xtab(ip1)-xtab(ix(ic)))
-          if(abs(dx).gt.1.e-20_r8)then
-             t(ic)=(x(ic)-xtab(ix(ic)))/dx
-          else
-             t(ic)=0._r8
-          endif
-	end do
-      else
-        ix(:ncol)=1
-        t(:ncol)=0._r8
-      endif
-      if(jm.gt.1)then
-        do ic=1,ncol
-          do j=1,jm
-            if(y(ic).lt.ytab(j))go to 20
-          enddo
-   20     jy(ic)=max0(j-1,1)
-          jp1=min(jy(ic)+1,jm)
-          dy=(ytab(jp1)-ytab(jy(ic)))
-          if(abs(dy).gt.1.e-20_r8)then
-             u(ic)=(y(ic)-ytab(jy(ic)))/dy
-             if(u(ic).lt.0._r8.or.u(ic).gt.1._r8)then
-                write(iulog,*) 'u,y,jy,ytab,dy=',u(ic),y(ic),jy(ic),ytab(jy(ic)),dy
-             endif
-          else
-            u(ic)=0._r8
-          endif
-	end do
-      else
-        jy(:ncol)=1
-        u(:ncol)=0._r8
-      endif
-   30 continue
-      do ic=1,ncol
-         tu(ic)=t(ic)*u(ic)
-         tuc(ic)=t(ic)-tu(ic)
-         tcuc(ic)=1._r8-tuc(ic)-u(ic)
-         tcu(ic)=u(ic)-tu(ic)
-         jp1=min(jy(ic)+1,jm)
-         ip1=min(ix(ic)+1,im)
-         do k=1,km
-            out(ic,k)=tcuc(ic)*table(k,ix(ic),jy(ic))+tuc(ic)*table(k,ip1,jy(ic))   &
-               +tu(ic)*table(k,ip1,jp1)+tcu(ic)*table(k,ix(ic),jp1)
-	 end do
-      enddo
-      return
-      end subroutine binterp
+
+subroutine binterp(table, ncol, ref_real, ref_img, ref_real_tab, ref_img_tab, &! in
+     itab, jtab, ttab, utab, coef) !inout/out
+
+  !------------------------------------------------------------------------------
+  ! Bilinear interpolation along the refractive index dimensions
+  ! of the table to estimate Chebyshev coefficients at an
+  ! intermediate refractive index.
+
+  ! In short wave, the first call computes itab, jtab, ttab, utab and coef.
+  ! The subsequent calls use itab, jtab, ttab and utab as inputs and compute coef
+
+  ! In long wave, we have just one call to compute itab,jtab,ttab, utab and coef
+  !------------------------------------------------------------------------------
+  implicit none
+
+  !intent-ins
+  integer, intent(in)  :: ncol
+  real(r8), intent(in) :: table(ncoef,prefr,prefi)
+  real(r8), intent(in) :: ref_real(pcols), ref_img(pcols) !real and imganinary parts of refractive indices [unitless]
+  real(r8), intent(in) :: ref_real_tab(prefr), ref_img_tab(prefi) !!real and imganinary table refractive indices [unitless]
+
+  !intent-inouts/outs
+  integer,  intent(inout) :: itab(pcols), jtab(pcols)
+  real(r8), intent(inout) :: ttab(pcols), utab(pcols)
+  real(r8), intent(out)   :: coef(pcols,ncoef) !coefficient interpolated bilinearly
+
+  !local
+  integer  :: ip1, jp1, icoef, ic
+  real(r8) :: tu(pcols), tuc(pcols),tcu(pcols), tcuc(pcols)
+
+  if(itab(1) <= 0) then
+
+     !compute factors for the real part
+     call compute_factors(prefr, ncol, ref_real, ref_real_tab, & !in
+          itab, ttab) !out
```

Comment:
this should be inout

Reply:
I ended up keeping it `out` following your other comment.